pkg/grpc/server.go

```diff
@@ -100,12 +94,21 @@ type API interface {
 	Start() *concurrency.Signal
 	// Register adds a new APIService to the list of API services
 	Register(services ...APIService)
+
+	// Stop will shutdown all listeners and stop the HTTP/gRPC multiplexed server. This process will run in the background
+	// and returns a signal that can be checked for when the shutdown finished.
+	Stop() *concurrency.Signal
 }
 
 type apiImpl struct {
 	apiServices        []APIService
 	config             Config
 	requestInfoHandler *requestinfo.Handler
+	listeners          []serverAndListener
+
+	grpcServer            *grpc.Server
+	grpcShutdownRequested concurrency.Signal
```

Comment:
```suggestion
	grpcShutdownInProgress concurrency.Signal
```

Reply:
Imho the easiest way I see is making the `Stop` call blocking, guarding it with a `mutex` and functions which need to check can check if a shutdown is in progress by checking the mutex state.

Reply:
Check latest code.